controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1355,6 +1357,86 @@ func (r *BareMetalHostReconciler) actionDeprovisioning(prov provisioner.Provisio
 	return actionComplete{}
 }
 
+func (r *BareMetalHostReconciler) doServiceIfNeeded(prov provisioner.Provisioner, info *reconcileInfo, hup *metal3api.HostUpdatePolicy) (result actionResult) {
+	servicingData := provisioner.ServicingData{}
+
+	// (NOTE)janders: since Servicing is an opt-in feature that requires HostUpdatePolicy to be created and set to onReboot
+	// set below booleans to false by default and change to true based on policy settings
+
+	var fwDirty bool
+	var hfsDirty bool
+	var liveFirmwareSettingsAllowed bool
+
+	if hup != nil {
+		liveFirmwareSettingsAllowed = (hup.Spec.FirmwareSettings == metal3api.HostUpdatePolicyOnReboot)
+	}
+
+	if liveFirmwareSettingsAllowed {
+		// handling pre-HFS FirmwareSettings here
+		if !reflect.DeepEqual(info.host.Status.Provisioning.Firmware, info.host.Spec.Firmware) {
+			servicingData.FirmwareConfig = info.host.Spec.Firmware
+			fwDirty = true
+		}
+		// handling HFS based FirmwareSettings here
+		var hfs *metal3api.HostFirmwareSettings
+		var err error
+		hfsDirty, hfs, err = r.getHostFirmwareSettings(info)
+		if err != nil {
+			return actionError{fmt.Errorf("could not determine updated settings: %w", err)}
+		}
+		if hfsDirty {
+			servicingData.ActualFirmwareSettings = hfs.Status.Settings
+			servicingData.TargetFirmwareSettings = hfs.Spec.Settings
+		}
+	}
+
+	dirty := fwDirty || hfsDirty
+
+	// Even if settings are clean, we need to check the result of the current servicing.
+	if !dirty && info.host.Status.OperationalStatus != metal3api.OperationalStatusServicing && info.host.Status.ErrorType != metal3api.ServicingError {
+		// If nothing is going on, return control to the power management.
+		return nil
+	}
+
+	// FIXME(janders/dtantsur): this implementation may lead to a scenario where if we never actually
+	// succeed before leaving this state (e.g. by deprovisioning) we lose the signal that the
+	// update didn't actually happen. This is deemed an acceptable risk for the moment since it is only
+	// going to impact a small subset of Firmware Settings implementations.
+	currentError := info.host.Status.ErrorType
+	if clearErrorWithStatus(info.host, metal3api.OperationalStatusServicing) {
```

Comment:
I think this is too early to clear an error. We should do it after the check for `provResult.ErrorMessage != ""` around line 1418.

It's also arguably too late for an attempt to set the status to Servicing in the non-error case, because we'll still only write it once servicing has already started.

Reply:
I changed the order as per the first line of the comment. Unsure about the second part - will check in with Dmitry.

Reply:
Remember that writes to the k8s API only happen after we return from this function. So if you want to put it into `OperationalStatusServicing` _before_ servicing starts then you'd need to do something like:

```
    if info.host.Status.OperationalStatus != metal3api.OperationalStatusServicing {
        info.host.Status.OperationalStatus = metal3api.OperationalStatusServicing
        return actionUpdate{}
    }
```

Just changing the order has no effect in that respect.

Reply:
If we do it this way, we'll lose the error information, and it will never be passed to `prov.Service`.

Reply:
I stand corrected. If we don't call `clearErrorWithStatus`, we'll have `ErrorType` still set. We just need to remember to unset it after the successful call.

Maybe we need to return `clearErrorWithStatus` to where it was before, but also set the status to servicing explicitly in the way that Zane suggested?